wlauto/instrumentation/servo_power_monitors/__init__.py

```diff
@@ -0,0 +1,179 @@
+        if not self.labels:
+            self.labels = ["PORT_{}".format(channel) for channel, _ in enumerate(self.power_domains)]
+        self.power_data = None
+        self.stopped = True
+
+    def validate(self):
+        if self.labels and not len(self.power_domains) == len(self.labels):
+            raise ConfigError('There should be exactly one label per power domain')
+
+    def setup(self, context):
+        # pylint: disable=access-member-before-definition
+        self.output_directory = os.path.join(context.output_directory, 'servo_power_monitor')
+        self.outfiles = [os.path.join(self.output_directory, '{}.csv'.format(label)) for label in self.labels]
+        self.poller = PowerPoller(self.cros_session, self.command, self.outfiles, self.sampling_rate)
+
+    def start(self, context):
+        self.poller.start()
+        self.stopped = False
+
+    def stop(self, context):
+        self.power_data = self.poller.stop()
+        self.poller.join()
+        self.stopped = True
+
+    def update_result(self, context):
+        os.mkdir(os.path.dirname(self.outfiles[0]))
+        for i, outfile in enumerate(self.outfiles):
+            self.power_data[i] = [float(v) / 1000.0 for v in self.power_data[i]]
+            metric_name = '{}_power'.format(self.labels[i])
+            sample_sum = sum(self.power_data[i])
+            power = sample_sum / len(self.power_data[i])
+            context.result.add_metric(metric_name, round(power, 3), 'Watts')
+            metric_name = '{}_energy'.format(self.labels[i])
+            energy = sample_sum * (1.0 / self.sampling_rate)
+            context.result.add_metric(metric_name, round(energy, 3), 'Joules')
+            with open(outfile, 'wb') as f:
+                c = csv.writer(f)
+                c.writerow(['{}_power'.format(self.labels[i])])
+                for val in self.power_data[i]:
+                    c.writerow([val])
+
+    def teardown(self, context):
+        if not self.stopped:
+            self.stop(context)
+        self.server_session.kill_session()
+        self.cros_session.kill_session()
+
+
+class PowerPoller(threading.Thread):
+
+    def __init__(self, cros_session, command, outputfiles, sampling_rate):
+        super(PowerPoller, self).__init__()
+        self.command = command
+        self.outputfiles = outputfiles
+        self.power_data = [[] for _ in outputfiles]
+        self.samples_per_frame = sampling_rate
+        self.period = 1.0 / sampling_rate
+        self.term_signal = threading.Event()
+        self.term_signal.set()
+        self.cros_session = cros_session
+
+    def run(self):
+        while self.term_signal.is_set():
+            time.sleep(self.period)
+            self.cros_session.send_command(self.command)
+            lines = self.cros_session.get_lines()
+            for i, line in enumerate(lines):
+                _, value = line.split(':')
+                self.power_data[i].append(value)
```

Comment:
This isnt time thats being discarded its the name of the power domain the reading is from e.g: